### PR TITLE
OP-6 new proposed pattern

### PR DESCRIPTION
**Pull request by mwithi**
New proposed pattern as for discussion outcome here

https://openhospital.atlassian.net/wiki/spaces/OHC/pages/769622123/Rest+API+Controller+pattern

TODO:
- fix tests
- verify and simplify accounting and admission mappers

### CODE REVIEW

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 1]
```diff
@@ -61,6 +61,9 @@
 	@Mock
     private PatientBrowserManager patientBrowserManagerMock;
 	
+	@Autowired
```
Comment (emecas) [inline src/test/java/org/isf/patient/rest/PatientControllerTest.java:64]: As i mentioned before this  tests suite  pretends to be Spring independent, you would need to remove `@Autowired ` , instance the mapper manually, and provide the mapper in the setup method to the controller constructor like this:

```

    private static PatientMapper mapper = new PatientMapper();
	
    @Before
    public void setup() {
    	MockitoAnnotations.initMocks(this);
    	this.mockMvc = MockMvcBuilders
				.standaloneSetup(new PatientController(patientBrowserManagerMock, mapper))
   				.setControllerAdvice(new OHResponseEntityExceptionHandler())
   				.build();
    }
```

File: src/main/java/org/isf/shared/GenericMapper.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.modelmapper.ModelMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class GenericMapper<SourceType, DestType> implements Mapper<SourceType, DestType> {
+
+	@Autowired
```
[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/shared/GenericMapper.java:12]: Hi @emecas ,thanks for this review. The @Autowired on the modelMapper is needed to use the  @Bean defined into ModelMapperConfig, so also with the proposed change in https://github.com/informatici/openhospital-api/pull/24 it doesn't work for me as reported in the related review.

So I guess we should build the mapper indipendently in order to run the test without Spring support, what do you think?

[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/shared/GenericMapper.java:12]: or to remove @Autowired from other properties in ModelMapperConfig like this:
```
@Configuration
public class ModelMapperConfig {

    protected BlobToByteArrayConverter blobToByteArrayConverter = new BlobToByteArrayConverter();

    protected ByteArrayToBlobConverter byteArrayToBlobConverter = new ByteArrayToBlobConverter();

    @Bean
    public ModelMapper modelMapper() {
        ModelMapper modelMapper = new ModelMapper();
        modelMapper.addConverter(blobToByteArrayConverter);
        modelMapper.addConverter(byteArrayToBlobConverter);
        return modelMapper;
    }
}
```

[the remark this answers is not in the document]
Comment (emecas) [reply src/main/java/org/isf/shared/GenericMapper.java:12]: @mwithi  I have added a fix on #24 handling that particular configuration issue. Let me know if that accomplishes your original idea for the new pattern?

[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/shared/GenericMapper.java:12]: Also I think you are using Spring:
```
@RunWith(SpringRunner.class)
@SpringBootTest
public class OpenHospitalApiApplicationTests {
	@Test
	public void contextLoads() {
	}
}
```
so every class should be in the context as well, no need to "new" anything, what do you think?

[the remark this answers is not in the document]
Comment (emecas) [reply src/main/java/org/isf/shared/GenericMapper.java:12]: That test was already there, let's keep it for checking the app is loading properly.

I am not using SpringRunner in the Test Suite proposal, you can check it by removing  the `new ModelMapper()` and executing `PatientControllerTest`,  it is going to give you `NullPointerException` for the tests using the mapper.

[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/shared/GenericMapper.java:12]: Yes, but what I mean, why to not use Spring and to have independent tests? In this way of course we have to rebuild all configurations apart, and in the end we are testing something that is different from the application context, what do you think about this?

[the remark this answers is not in the document]
Comment (emecas) [reply src/main/java/org/isf/shared/GenericMapper.java:12]: It depends what testing we want to perform, proposed test suite is focused on unit testing controllers methods in isolation. 

I feel the task scope for unit testing all controllers is enough big and complex for now.

Once unit testing is covered, we can start focusing on integration testing  subjects like the full API, data base,  security, resources, devices, etc.

By the time this new pattern is validated,  discused and adopted. I  could provide integration testing as well in a different task.

[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/shared/GenericMapper.java:12]: I agree with the roadmap, nevertheless unit test doesn't mean that we cannot use Spring, isolation can be granted within Spring I think, we should create a specific context for the tests...

**Comment by mwithi** [comment]
According with the yesterday online meeting I think we agreed to merge this PR as is.
@vicusotto @npasquetto @averni @gildasdongmo are you ok with it?

**Comment by gildasdongmo** [comment]
> According with the yesterday online meeting I think we agreed to merge this PR as is.
> @vicusotto @npasquetto @averni @gildasdongmo are you ok with it?

It is ok for me.

**Comment by mwithi** [comment]
Hi @vicusotto, thanks for the review. I know that could be a temporary workaround, but still haveing PatientMapper not autowired, are you sure it will work fine with its configuration (ModelMapperConfig)? I think it will not save the patient photo in that way... did you test it with Swagger?

**Comment by vicusotto** [comment]
Hi @mwithi ,
Yes it works. It's how I use Spring Autowired with Constructor-based dependency injection. 
I put Autowired on Constructor instead on fields.
So it's not related with the fix but only my best practice.
Since all project works with field base annotation I update the #25  so it's aligned with other controllers.

The fix consist to instantiate modelMapper with Relfection.
So, it's the same of #24 @emecas fix proposal without modyfing GenericMapper's Constructor.

`ReflectionTestUtils.setField(mapper, "modelMapper", modelMapper);`

**Comment by mwithi** [comment]
> I put Autowired on Constructor instead on fields.

Sorry, I didn't notice that! Ok, I think it should work in this way.
Thanks!

**Comment by emecas** [comment]
> @emecas does it work for you? do you agree to close #24 in favor of #23?

sure @mwithi,  if that is the most convenient for you, actually providing #24 was more about making available a solution for the tests.

On the other hand, I  just try #23 and it is still failing,  I am not sure if i could be missing any thing  else, but the are 9 tests of the 18 in `PatientControllerTest` are failing with a:

```
java.lang.NullPointerException
	at org.isf.patient.rest.PatientControllerTest$PatientDTOHelper.setup(PatientControllerTest.java:567)
	at org.isf.patient.rest.PatientControllerTest.when_post_patients_PatientBrowserManager_getPatient_returns_null_then_OHAPIException_BadRequest(PatientControllerTest.java:134)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
```

I am running on your 060ec3ea3 last commit here and hitting 2 different commits on core master 69627f8 and 6366099 with same results for both cases. 

Here you have the details: 

image

Thanks @vicusotto

**Comment by vicusotto** [comment]
Hi @mwithi and @emecas ,

> @emecas does it work for you? do you agree to close #24 in favor of #23?

Maybe you meant #25 instead of #24 ?

> On the other hand, I just try #23 and it is still failing, I am not sure if i could be missing any thing else, but the are 9 tests of the 18 in PatientControllerTest are failing with a:

In #25 you should not have NPE failure.

> Not only, you introduced also the model mapper configuration in the test setup, that is ok, and it is what I meant with "we have to rebuild all configurations apart", but as a temporary workaround can be ok.

Ok!

Thanks!

**Comment by emecas** [comment]
> Hi @mwithi and @emecas ,
> 
> > @emecas does it work for you? do you agree to close #24 in favor of #23?
> 
> Maybe you meant #25 instead of #24 ?
> 

Yes @vicusotto I think so I did think your fixes were already in #23. So  moving further I reviewed #25 and added a PR to your repo in order to express clearly what i meant.  Thanks again

**Comment by vicusotto** [comment]
Hi @emecas ,
thank you for your review and your suggestion for PR #25 . 
I agree with you and merge your PR cleaning solution.

Thanks!

**Comment by mwithi** [comment]
Thank you guys! So if I'm not wrong we are ager to merge https://github.com/informatici/openhospital-api/pull/25 and then https://github.com/informatici/openhospital-api/pull/23, right?

**Comment by vicusotto** [comment]
Hi @mwithi ,
yes, I agree.

Thanks!

**Pull request merged**